### PR TITLE
base types for events package

### PR DESCRIPTION
**Pull request by jimlambrt**
This is first of 3 PRs to add the events package and they should be reviewed in order.  After this one, then #1276 and #1277.

This PR will be merged into main after the last one in the series (#1277) has been approved.

BTW, none of these 3 PRs will affect any of the rest of Boundary.  They are stand-alone infrastructure for eventing audits, errors, and observations.  Subsequent PRs may change how some of this infrastructure works and we work to integrating the events package with the rest of Boundary and we continue to receive feedback from the Event Logging RFCs.

### CODE REVIEW

File: internal/observability/events/event_type.go
Comment (s-christoff) [inline internal/observability/events/event_type.go:6]: Opinions on changing EveryType to AllType(s)?

**Review by s-christoff — DISMISSED** [review]
Some minor nits but nothing huge - thanks for laying this foundation :)

File: internal/observability/events/event.go
Comment (jimlambrt) [reply internal/observability/events/event.go:18]: Agreed. I was sort of holding off a bit on documenting them until your RFC was more ratified (which includes these structs).  Once we have the "royal" blessing, we can open a PR to spruce up some godoc comments for these.

File: internal/observability/events/event_audit.go
Comment (jimlambrt) [reply internal/observability/events/event_audit.go:33]: The entry point for writing Audit events will be: event.WriteAudit(ctx context.Context, a *Audit, opt...Option)  The Audit struct has to be exported since it's referenced in the exported WriteAudit(...), but newAudit(...) should never be called by anyone outside the event package.

File: internal/observability/events/event_audit.go
Comment (jimlambrt) [reply internal/observability/events/event_audit.go:45]: datetime

File: internal/observability/events/event.go
Comment (mgaffney) [inline internal/observability/events/event.go:1]: The package name and directory name do not match.

File: internal/observability/events/event_type.go
Comment (jimlambrt) [reply internal/observability/events/event_type.go:6]: I think EveryType (singular) reads better than AllType (singular) and all the type definitions are singulars.

File: internal/observability/events/eventer.go
Comment (mgaffney) [inline internal/observability/events/eventer.go:4]: Are the comments for each `const` field correct?

File: internal/observability/events/options.go
Comment (mgaffney) [inline internal/observability/events/options.go:66]: `event.WithNow` reads odd to me.

**Review by mgaffney — CHANGES_REQUESTED** [review]
Many (most?) of the exported types, functions, constant, etc are missing docs.

File: internal/observability/events/eventer.go
Comment (mgaffney) [reply internal/observability/events/eventer.go:4]: I misread the comment. Please ignore.

File: internal/observability/events/options.go
Comment (s-christoff) [reply internal/observability/events/options.go:66]: +1 to this

Comment (jimlambrt) [reply internal/observability/events/options.go:66]: Hum... perhaps.  BTW, it's only used in tests, but had to be exported since some of those tests will not be in the event package but rather in event_test

**Comment by jimlambrt** [comment]
> Many (most?) of the exported types, functions, constant, etc are missing docs.
this will be addressed in a future PR. See my previous comment: https://github.com/hashicorp/boundary/pull/1275#discussion_r645734666

File: internal/observability/events/event_audit.go
Comment (vancluever) [inline internal/observability/events/event_audit.go]: Curious what the intent here is for using `sync.Once` to initialize this - only thing I can think of is to protect against concurrency, but I don't see anything here that would need that? Would a simple empty check/non-empty assertion on `gated.Id` be better?

```go
if gated.Id == "" {
  return "", nil, errors.New(errors.InvalidParameter, op, fmt.Sprintf("event %d: Id is required", i))
}

if validId == {
  validId = gated.Id
}
```

File: internal/observability/events/options.go
Comment (jimlambrt) [reply internal/observability/events/options.go:66]: `event` is the package name, so not sure that can be changed.  The prefix `With` is our standard for options and `Now` aligns with the std lib `time.Now()` and every other instance of a time.Time that represents now.

File: internal/observability/events/event_audit_test.go
Comment (vancluever) [inline internal/observability/events/event_audit_test.go]: Should we type out `Audit.Type` so that this kind of type conversion is not necessary? Assuming this field will always be non-arbitrary.

**Comment by jimlambrt** [comment]
@mgaffney @s-christoff added a commit with godoc comments for exported fields except for the unratified audit related fields and/or structs.  Ty for the suggestion!

File: internal/observability/events/event.go
Comment (jimlambrt) [reply internal/observability/events/event.go]: fixed. ty!

File: internal/observability/events/event.go
Comment (jimlambrt) [reply internal/observability/events/event.go]: @s-christoff I've pushed a commit with additional godoc comments for the exported bits with the exception of unratified audit fields and/or structs.

File: internal/observability/events/event_error_test.go
Comment (vancluever) [inline internal/observability/events/event_error_test.go]: _Minor_ just noting that I had to actually check here to see that `wantErr` was actually the valid case here, and not error validation. Might be worth just changing this to `want`?

File: internal/observability/events/event_observation_test.go
Comment (vancluever) [inline internal/observability/events/event_observation_test.go]: This field is never referenced in any of the tests, so the `valid` test is not being asserted, for example.

File: internal/observability/events/options.go
Comment (vancluever) [reply internal/observability/events/options.go]: I think `WithTime` would be better? Given that it takes a `time.Time` and does not generate one via `time.Now()`.

**Review by vancluever — CHANGES_REQUESTED** [review]
Agreeing with @mgaffney's points for sure re: comments and package naming (although some of these could be addressed now as I have not hit "refresh" yet 😜 )

Also some other things that need fixing including a test where an assertion seems to be getting missed.

File: internal/observability/event/options.go
Comment (louisruch) [inline internal/observability/event/options.go:21]: Perhaps `withId Id`

**Review by louisruch — COMMENTED** [review]
This looks great, I have a couple questions and comments, I see these PR's are getting some traction on reviews - I read through previous comments sorry in advance if I doubled up on any comment/question

File: internal/observability/event/id.go
Comment (louisruch) [reply internal/observability/event/id.go:8]: I see you are using `newId` in follow up PRs outside the scope of Type it seems

File: internal/observability/events/event_audit_test.go
Comment (jimlambrt) [reply internal/observability/events/event_audit_test.go:33]: Audit.Type is string based on an effort to adhere to the Audit schema put forth on ENG-030 and CSL-121.  I used a defined type internally because it afford some advantages... but we should adhere to the previous work by the Nomad and Consul teams when it comes to the outbound audit event data.

File: internal/observability/events/event_error_test.go
Comment (jimlambrt) [reply internal/observability/events/event_error_test.go:95]: ty!

File: go.mod
Comment (jimlambrt) [reply go.mod]: ty!

File: internal/observability/event/event_audit.go
Comment (jimlambrt) [reply internal/observability/event/event_audit.go]: ty!

File: internal/observability/events/event_observation_test.go
Comment (jimlambrt) [reply internal/observability/events/event_observation_test.go:102]: yep... I went with wantErrMatch and forgot to remove it.

ty!

File: internal/observability/event/event_audit.go
Comment (jimlambrt) [reply internal/observability/event/event_audit.go]: the last event will win by design.

File: internal/observability/events/event_audit.go
Comment (jimlambrt) [reply internal/observability/events/event_audit.go]: ty!

**Review by louisruch — DISMISSED** [review]
lgtm

**Comment by jimlambrt** [comment]
I've rolled back the directory name change for now... it will be reapplied before merging but it was causing all sorts of conflicts with the upstream PRs

**Review by vancluever — APPROVED** [review]
:shipit: 

Just want to bring up as a minor point https://github.com/hashicorp/boundary/pull/1275/files#r645785137 (the thread over `WithNow` and how it might be better as `WithTime` since this option is not actually deriving or enforcing a current time.

**Pull request merged**
